internal/discovery/discovery.go

```diff
@@ -239,21 +235,8 @@ func (r *CRDiscoverer) PollForCacheUpdates(
 		r.SafeWrite(func() {
 			r.WasUpdated = false
 		})
-		// Run the metrics handler with updated configs.
-		olderContext, olderCancel = context.WithCancel(ctx)
-		go func() {
-			// Blocks indefinitely until the unbuffered context is cancelled to serve metrics for that duration.
-			err = m.Run(olderContext)
-			if err != nil {
-				// Check if context was cancelled.
-				select {
-				case <-olderContext.Done():
-					// Context cancelled, don't really need to log this though.
-				default:
-					klog.ErrorS(err, "failed to run metrics handler")
-				}
-			}
-		}()
+		// Update metric handler with the new configs.
+		m.ReconfigureSharding(ctx)
```

Comment:
I don't think we need this do we? None of the values set when reconfiguring sharing are changing when a new CRD is added or removed and because we don't cancel the metrics handler anymore here, it shouldn't impact sharding at all.

Reply:
I think after [these changes above](https://github.com/kubernetes/kube-state-metrics/blob/4aced25f69e00d54ec3e03a921f8bf60d4e8a783/internal/discovery/discovery.go#L224-L232) in the `storeBuilder`, `configureSharding` has to build `metricsHandler` again, [here](https://github.com/kubernetes/kube-state-metrics/blob/4aced25f69e00d54ec3e03a921f8bf60d4e8a783/pkg/metricshandler/metrics_handler.go#L99), because some values like `storeBuilder.enabledResources` have changed

Reply:
You are correct, we need to rebuild the writers and more importantly the underlying stores. But we don't have to touch the number of shards.

But do you think you could decouple https://github.com/kubernetes/kube-state-metrics/blob/4aced25f69e00d54ec3e03a921f8bf60d4e8a783/pkg/metricshandler/metrics_handler.go#L99 from `ConfigureSharding` and put it in its own method like `BuildWriters` or something?

The reasoning is that we now need to reconfigure the writers, when a CRD is created/deleted and the context is cancelled.

But there is still something that annoys me here, when we detect changes to CRDs, we shouldn't have to rebuild everything. I'll continue to look into it and see how we could address that, but in the meantime I am fine with just addressing the issue at hand.

Reply:
Agreed. It can be confusing to have a reference to shards when they didn't change. Added 👍